CRM/MembershipExtras/Hook/Pre/Contribution.php

```diff
@@ -0,0 +1,108 @@
+
+class CRM_MembershipExtras_Hook_Pre_Contribution {
+
+
+  private $id;
+
+  /**
+   * The contribution to-be-created parameters passed from the hook.
+   *
+   * @var array
+   */
+  private $params;
+
+  /**
+   * The payment plan recurring contribution data if exist
+   *
+   * @var array|null
+   */
+  private $recurContribution = NULL;
+
+  /**
+   * Is this an offline payment plan payment ?
+   *
+   * @var bool
+   */
+  private $isOfflinePaymentPlanPayment = FALSE;
+
+  public function __construct($id, &$params) {
+    $this->id = $id;
+    $this->params = $params;
+    $this->setRecurContribution();
+    $this->setIsOfflinePaymentPlanPayment();
+  }
+
+  /**
+   * Sets $recurContribution
+   */
+  private function setRecurContribution() {
+    if (empty($this->params['prevContribution']->contribution_recur_id)) {
+      return;
+    }
+
+    $this->recurContribution = civicrm_api3('ContributionRecur', 'get', [
+      'sequential' => 1,
+      'id' => $this->params['prevContribution']->contribution_recur_id,
+    ])['values'][0];
+  }
+
+  /**
+   * Sets $isOfflinePaymentPlanPayment
+   */
+  private function setIsOfflinePaymentPlanPayment() {
+    $contributionStatuses = CRM_Contribute_PseudoConstant::contributionStatus(NULL, 'name');
+    $previousStatus = $contributionStatuses[$this->params['prevContribution']->contribution_status_id];
+    $currentStatus = $contributionStatuses[$this->params['contribution_status_id']];
+
+    if ($previousStatus === 'Pending' && in_array($currentStatus, ['Partially paid', 'Completed']) && !empty($this->recurContribution)) {
+      $manualPaymentProcessors = CRM_MembershipExtras_Service_ManualPaymentProcessors::getIDs();
+      $isOfflineContribution = empty($this->recurContribution['payment_processor_id']) ||
+        in_array($this->recurContribution['payment_processor_id'], $manualPaymentProcessors);
+
+      if (!empty($this->recurContribution['installments']) || !$isOfflineContribution) {
+        $this->isOfflinePaymentPlanPayment = TRUE;
+      }
+    }
+  }
+
+  /**
+   * Updates the payment plan recurring contribution
+   * status based on the number of completed/partially paid
+   * installments.
+   */
+  public function updatePaymentPlanStatus() {
+    if (!$this->isOfflinePaymentPlanPayment) {
+      return;
+    }
+
+    $paidInstallmentsCount = civicrm_api3('Contribution', 'getcount', [
+      'contribution_recur_id' => $this->recurContribution['id'],
+      'contribution_status_id' => 'Completed',
+    ]);
+
+    $partiallyPaidInstallmentsCount = civicrm_api3('Contribution', 'getcount', [
+      'contribution_recur_id' => $this->recurContribution['id'],
+      'contribution_status_id' => 'Partially paid',
+    ]);
+
+    $newStatus = NULL;
+    if (($partiallyPaidInstallmentsCount == 1 && $paidInstallmentsCount == 0) ||
+      ($partiallyPaidInstallmentsCount == 0 && $paidInstallmentsCount == 1) ||
+      ($partiallyPaidInstallmentsCount == 0 && $paidInstallmentsCount == 0)
+    ) {
```

Comment:
Shouldn't the recurring contribution be 'In Progress' if either of those values is more than zero and paid installments is less than the number of installments? Although, since 'Completed' status is taken care of on the next if statement, it might be enough just to check if either of those values is more than zero. Also, I don't really understand why if both are zero, the recurring contribution would be In progress... shouldn't it stay 'Pending' if no payments have started?

Eg.
```php
if ($partiallyPaidInstallmentsCount > 0 || $paidInstallmentsCount > 0) {
  $newStatus = 'In Progress';
}
```